File: packages/core/src/services/query/query.ts
import { QueryHash, Incantation as IncantationData, QueryTimeToLive } from '../../types.js';
import { Table, RecordId, Duration, Uuid } from 'surrealdb';
// import { RemoteDatabaseService } from '../database/remote.js'; // REMOVED
import { LocalDatabaseService } from '../database/local.js';
import { Incantation } from './incantation.js';
import { createLogger, Logger } from '../logger/index.js';
import {
  createQueryEventSystem,
  QueryEventSystem,
  QueryEventTypeMap,
  QueryEventTypes,
} from './events.js';
import { Event } from '../../events/index.js';
import { decodeFromSpooky, parseRecordIdString } from '../utils/index.js';
import { SchemaStructure, TableModel } from '@spooky/query-builder';
import { StreamProcessorService } from '../stream-processor/index.js';

const withRetry =
  (logger: Logger) =>
  async <T>(operation: () => Promise<T>, retries = 3, delayMs = 100): Promise<T> => {
    let lastError;
    for (let i = 0; i < retries; i++) {
      try {
        return await operation();
      } catch (err: any) {
        lastError = err;
        // Check for transaction error or generic connection issues
        if (
          err?.message?.includes('Can not open transaction') ||
          err?.message?.includes('transaction')
        ) {
          logger.warn(
            {
              attempt: i + 1,
              retries,
              error: err.message,
            },
            'Retrying DB operation due to transaction error'
          );

          await new Promise((res) => setTimeout(res, delayMs * (i + 1))); // Linear backoff
          continue;
        }
        throw err;
      }
    }
    throw lastError;
  };

export class QueryManager<S extends SchemaStructure> {
  private activeQueries: Map<QueryHash, Incantation<any>> = new Map();
  private liveQueryUuid: string | null = null;
  private events: QueryEventSystem;
  private logger: Logger;

  public get eventsSystem() {
    return this.events;
  }

  constructor(
    private schema: S,
    private local: LocalDatabaseService,
    private streamProcessor: StreamProcessorService,
    private clientId: string | undefined, // undefined is valid for optional clientId, but argument position is fixed
    logger: Logger
  ) {
    this.logger = logger.child({ service: 'QueryManager' });
    this.events = createQueryEventSystem();
    this.events.subscribe(
      QueryEventTypes.IncantationIncomingRemoteUpdate,
      this.handleIncomingUpdate.bind(this)
    );
  }

  public getQueriesThatInvolveTable(tableName: string) {
    return [...this.activeQueries.values()].filter((q) => q.invlovesTable(tableName));
  }

  public getActiveQueries() {
    return Array.from(this.activeQueries.values());
  }

  public async init() {
    // ClientId setup moved to SpookySync/Auth
  }

  public handleIncomingUpdate(
    eventOrPayload: Event<QueryEventTypeMap, 'QUERY_INCANTATION_INCOMING_REMOTE_UPDATE'> | any
  ) {
    const payload =
      eventOrPayload && 'payload' in eventOrPayload ? eventOrPayload.payload : eventOrPayload;
    const { incantationId, records = [], remoteHash, remoteArray } = payload;

    if (!incantationId || !incantationId.id) {
      this.logger.error({ payload }, '[QueryManager] Invalid payload: missing incantationId');
      return;
    }

    const incantation = this.activeQueries.get(incantationId.id.toString());
    if (!incantation) {
      this.logger.warn(
        { incantationId: incantationId.toString() },
        '[QueryManager] Incantation not found for update'
      );
      return;
    }

    const validRecords: any[] = [];

    for (const r of records) {
      const decoded = decodeFromSpooky(
        this.schema,
        incantation.tableName,
        r as unknown as TableModel<
          Extract<
            S['tables'][number],
            {
              name: string;
            }
          >
        >
      );
      validRecords.push(decoded);
    }

    this.logger.debug(
      {
        incantationId: incantationId.toString(),
        queryHash: incantationId.id.toString(),
        totalRecords: records.length,
        validRecords: validRecords.length,
      },
      'Handling incoming remote update'
    );
    this.logger.debug(
      {
        incantationId: incantationId?.toString(),
        remoteHash,
        remoteArray: remoteArray ? 'PRESENT' : 'MISSING',
        localHashInPayload: (payload as any).localHash,
        localArrayInPayload: (payload as any).localArray ? 'PRESENT' : 'MISSING',
      },
      '[QueryManager] handleIncomingUpdate payload'
    );

    incantation.updateLocalState(validRecords, incantation.localHash, incantation.localArray);
    // Explicitly update remote state
    incantation.remoteHash = remoteHash;
    incantation.remoteArray = remoteArray;
    this.events.emit(QueryEventTypes.IncantationUpdated, {
      incantationId,
      records: validRecords,
      localHash: incantation.localHash,
      localArray: incantation.localArray,
      remoteHash: remoteHash,
      remoteArray: remoteArray,
    });

    // Verification and purging of orphans is now handled by SpookySync
  }

  public async handleStreamUpdate(update: any) {
    const { query_id, localHash, result_ids, localArray } = update;

    // query_id from StreamProcessor is a string (e.g. "_spooky_incantation:...")
    // We need to convert it to a RecordId for the event payload
    let incantationRecordId: RecordId;
    try {
      incantationRecordId = parseRecordIdString(query_id);
    } catch (e) {
      // Fallback for bare IDs or testing
      incantationRecordId = new RecordId('_spooky_incantation', query_id);
    }

    this.logger.debug(
      { incantationId: query_id, count: result_ids?.length },
      'Handling stream update'
    );

    if (!result_ids || result_ids.length === 0) {
      this.events.emit(QueryEventTypes.IncantationUpdated, {
        incantationId: incantationRecordId,
        records: [],
      });
      return;
    }

    try {
      const incantation = this.activeQueries.get(query_id);
      if (!incantation) {
        throw new Error('Incantation not found');
      }

      const [records] = await this.local.query<[Record<string, any>[]]>(
        incantation.surrealql,
        incantation.params
      );
      incantation.updateLocalState(records || [], localHash, localArray);

      this.events.emit(QueryEventTypes.IncantationUpdated, {
        incantationId: incantationRecordId,
        records: records || [],
        localHash,
        localArray,
      });
    } catch (err) {
      this.logger.error(
        { err, incantationId: query_id },
        'Failed to fetch records for stream update'
      );
    }
  }

  private async register(
    tableName: string,
    surrealql: string,
    params: Record<string, any>,
    ttl: QueryTimeToLive,
    involvedTables: string[] = []
  ): Promise<QueryHash> {
    this.logger.debug(
      {
        clientId: this.clientId,
        surrealql,
        params,
        ttl,
        involvedTables,
      },
      '[QueryManager] Registering incantation'
    );
    const id = await this.calculateHash({
      clientId: this.clientId,
      surrealql,
      params,
    });

    const recordId = new RecordId('_spooky_incantation', id);

    const existing = await this.registerStoreIncantation({
      recordId,
      surrealql,
      params,
      ttl,
      tableName,
      involvedTables,
    });

    if (!this.activeQueries.has(id)) {
      const incantation = new Incantation({
        id: recordId,
        surrealql,
        params,
        localHash: existing.localHash,
        localArray: existing.localArray,
        remoteHash: existing.remoteHash,
        remoteArray: existing.remoteArray,
        lastActiveAt: existing.lastActiveAt,
        ttl: existing.ttl,
        meta: {
          tableName,
          involvedTables,
        },
      });
      this.activeQueries.set(id, incantation);
      await this.initLifecycle(incantation);
    }
    return id;
  }

  async registerStoreIncantation({
    recordId,
    surrealql,
    params,
    ttl,
    tableName,
    involvedTables,
  }: {
    recordId: RecordId;
    surrealql: string;
    params: Record<string, any>;
    ttl: QueryTimeToLive;
    tableName: string;
    involvedTables: string[];
  }) {
    let [existing] = await withRetry(this.logger)(() =>
      this.local.query<[IncantationData]>('SELECT * FROM ONLY $id', {
        id: recordId,
      })
    );

    if (!existing) {
      existing = await withRetry(this.logger)(() =>
        this.local
          .getClient()
          .create<IncantationData>(recordId)
          .content({
            id: recordId,
            surrealQL: surrealql,
            params: params,
            clientId: this.clientId,
            localHash: '',
            localArray: [],
            remoteHash: '',
            remoteArray: [],
            lastActiveAt: new Date(),
            ttl: new Duration(ttl),
            meta: {
              tableName,
              involvedTables,
            },
          })
      );
    }

    if (!existing) {
      throw new Error('Failed to create or retrieve incantation');
    }
    return existing;
  }

  async query(
    tableName: string,
    surrealql: string,
    params: Record<string, any>,
    ttl: QueryTimeToLive,
    involvedTables: string[] = []
  ): Promise<QueryHash> {
    return this.register(tableName, surrealql, params, ttl, involvedTables);
  }

  private async initLifecycle(incantation: Incantation<any>) {
    const update = this.streamProcessor.registerIncantation(incantation);
    if (!update) {
      throw new Error('Failed to register incantation');
    }

    // Load cached records immediately to avoid UI flicker
    let records: Record<string, any>[] = [];
    try {
      const [result] = await this.local.query<[Record<string, any>[]]>(
        incantation.surrealql,
        incantation.params
      );
      records = result || [];
    } catch (err) {
      this.logger.warn({ err }, 'Failed to load initial cached records');
    }

    incantation.updateLocalState(records, update.localHash, update.localArray);

    this.events.emit(QueryEventTypes.IncantationInitialized, {
      incantationId: incantation.id,
      surrealql: incantation.surrealql,
      params: incantation.params ?? {},
      ttl: incantation.ttl,
      localHash: incantation.localHash,
      localArray: incantation.localArray,
      records,
    });

    await incantation.startTTLHeartbeat(() => {
      this.events.emit(QueryEventTypes.IncantationTTLHeartbeat, {
        incantationId: incantation.id,
      });
    });
  }

  subscribe(
    queryHash: string,
    callback: (records: Record<string, any>[]) => void,
    options: { immediate?: boolean } = {}
  ): () => void {
    const id = this.events.subscribe(QueryEventTypes.IncantationUpdated, (event) => {
      const incomingId = event.payload.incantationId.id.toString();
      if (incomingId === queryHash) {
        this.logger.debug(
          {
            queryHash,
            recordCount: event.payload.records.length,
          },
          'Subscription callback triggered'
        );

        callback(event.payload.records);
      } else {
        // this.logger.trace({ incomingId, queryHash }, 'Subscription ignored mismatch');
      }
    });

    if (options.immediate) {
      const records = this.activeQueries.get(queryHash)?.records ?? [];
      callback(records);
    }

    return () => {
      this.events.unsubscribe(id);
    };
  }

  private async calculateHash(data: any): Promise<string> {
    const content = JSON.stringify(data);
    const msgBuffer = new TextEncoder().encode(content);
    const hashBuffer = await crypto.subtle.digest('SHA-256', msgBuffer);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
  }
}

File: packages/flutter_core/lib/core/services/query/query.dart
import 'dart:async';
import 'dart:convert';

import '../database/local.dart';
import '../database/remote.dart';
import '../../types.dart' hide Incantation;
import '../events/main.dart'; // EventSystem
import '../sync/utils.dart'; // flattenIdTree

import 'package:flutter_surrealdb_engine/flutter_surrealdb_engine.dart';

import 'event.dart';
import 'incantation.dart';
import 'utils.dart'; // extractResult, parseRecordIdString
import '../mutation/events.dart'; // MutationEvent

class QueryManager {
  // Active queries map: hash -> Incantation
  final Map<String, Incantation<dynamic>> activeQueries = {};

  // Custom Event System for Query Events
  late final EventSystem<QueryEvent> events;

  final LocalDatabaseService local;
  final RemoteDatabaseService remote;
  final String? clientId;
  final bool enableLiveQuery;
  final EventSystem<dynamic> mutationEvents;

  // Constructor
  QueryManager({
    required this.local,
    required this.remote,
    required this.mutationEvents,
    this.clientId,
    this.enableLiveQuery = true,
  }) {
    events = EventSystem<QueryEvent>();
    // Subscribe to incoming remote updates
    events.subscribe<IncantationIncomingRemoteUpdate>(
      _handleIncomingRemoteUpdate,
    );
    // Subscribe to local mutations
    mutationEvents.subscribe<MutationEvent>(_handleLocalMutation);
  }

  // --- Public Methods ---

  Future<void> init() async {
    if (enableLiveQuery) {
      await _startLiveQuery();
    }
  }

  /// Register a new query/incantation.
  /// Returns the Hash ID of the query.
  Future<String> register({
    required String tableName,
    required String surrealql,
    required Map<String, dynamic> params,
    required QueryTimeToLive ttl,
  }) async {
    final id = await _calculateHash(surrealql, params);

    // Persist to local _spooky_incantation table
    final recordId = '_spooky_incantation:$id';

    // Prepare content
    final content = {
      'Id': id,
      'SurrealQL': surrealql,
      'Params': params,
      'ClientId': clientId,
      'Hash': id,
      'Tree': null,
      'LastActiveAt': DateTime.now().toIso8601String(),
      'TTL': ttl.value,
    };

    await local.query(
      'UPSERT $recordId CONTENT \$data',
      vars: {'data': content},
    );

    // Add to active memory map
    if (!activeQueries.containsKey(id)) {
      final incantation = Incantation(
        id: recordId,
        surrealql: surrealql,
        hash: id,
        ttl: ttl,
        meta: IncantationMeta(tableName: tableName),
        params: params,
        tree: null,
        lastActiveAt: DateTime.now(),
      );

      activeQueries[id] = incantation;
      await _initLifecycle(incantation);
    }

    return id;
  }

  /// Alias for register
  Future<String> query({
    required String tableName,
    required String surrealql,
    required Map<String, dynamic> params,
    required QueryTimeToLive ttl,
  }) {
    return register(
      tableName: tableName,
      surrealql: surrealql,
      params: params,
      ttl: ttl,
    );
  }

  /// Subscribe to changes for a specific query hash.
  /// Returns a cancel function.
  void Function() subscribe(
    String queryHash,
    void Function(List<Map<String, dynamic>> records) callback, {
    bool immediate = false,
  }) {
    final subscription = events.subscribe<IncantationUpdated>((event) {
      if (event.payload.incantationId == '_spooky_incantation:$queryHash') {
        callback(event.payload.records);
      }
    });

    if (immediate) {
      final incantation = activeQueries[queryHash];
      if (incantation != null && incantation.records != null) {
        final records = incantation.records!.cast<Map<String, dynamic>>();
        callback(records);
      } else {
        callback([]);
      }
    }

    return () => subscription.cancel();
  }

  // --- Internals ---

  Future<void> _initLifecycle(Incantation incantation) async {
    events.addEvent(
      IncantationInitialized(
        IncantationInitializedPayload(
          incantationId: incantation.id,
          surrealql: incantation.surrealql,
          ttl: incantation.ttl,
          params: incantation.params,
        ),
      ),
    );

    incantation.startTTLHeartbeat(() {
      events.addEvent(
        IncantationTTLHeartbeat(
          IncantationTTLHeartbeatPayload(incantationId: incantation.id),
        ),
      );
    });
  }

  void _handleIncomingRemoteUpdate(IncantationIncomingRemoteUpdate event) {
    final payload = event.payload;
    final parts = payload.incantationId.split(':');
    final queryHash = parts.length > 1 ? parts[1] : parts[0];

    final incantation = activeQueries[queryHash];
    if (incantation == null) return;

    final records = payload.records;

    IdTree? remoteTreeObj;
    if (payload.remoteTree is IdTree) {
      remoteTreeObj = payload.remoteTree as IdTree;
    } else if (payload.remoteTree != null) {
      try {
        remoteTreeObj = IdTree.fromJson(payload.remoteTree);
      } catch (e) {}
    }

    final remoteLeaves = flattenIdTree(remoteTreeObj);
    final remoteIds = remoteLeaves.map((l) => l.id).toSet();

    final validRecords = <Map<String, dynamic>>[];
    final orphanedRecords = <Map<String, dynamic>>[];

    for (final record in records) {
      final id = record['id'];
      final idStr = id.toString();

      if (remoteIds.contains(idStr)) {
        validRecords.add(record);
      } else {
        orphanedRecords.add(record);
      }
    }

    incantation.updateLocalState(
      validRecords,
      payload.remoteHash,
      payload.remoteTree,
    );

    events.addEvent(
      IncantationUpdated(
        IncantationUpdatedPayload(
          incantationId: payload.incantationId,
          records: validRecords,
        ),
      ),
    );

    if (orphanedRecords.isNotEmpty) {
      _verifyAndPurgeOrphans(orphanedRecords);
    }
  }

  Future<void> _handleLocalMutation(MutationEvent event) async {
    final affectedTables = <String>{};
    for (final mutation in event.payload) {
      final id = mutation.record_id;
      if (id != null) {
        final parts = id.split(':');
        if (parts.isNotEmpty) affectedTables.add(parts[0]);
      }
    }

    if (affectedTables.isEmpty) return;

    for (final incantation in activeQueries.values) {
      final table = incantation.meta?.tableName;
      if (table != null && affectedTables.contains(table)) {
        await _refreshLocalIncantation(incantation);
      }
    }
  }

  Future<void> _refreshLocalIncantation(Incantation incantation) async {
    try {
      final resStr = await local.query(
        incantation.surrealql,
        vars: incantation.params,
      );
      final dynamic raw = extractResult(resStr);
      List<Map<String, dynamic>> records = [];
      if (raw is List) {
        records = raw.cast<Map<String, dynamic>>();
      }

      incantation.records = records;

      events.addEvent(
        IncantationUpdated(
          IncantationUpdatedPayload(
            incantationId: incantation.id,
            records: records,
          ),
        ),
      );
    } catch (e) {
      print('[QueryManager] Failed to refresh local incantation: $e');
    }
  }

  Future<void> _verifyAndPurgeOrphans(
    List<Map<String, dynamic>> orphans,
  ) async {
    if (orphans.isEmpty) return;

    final idsToCheck = orphans.map((r) {
      final idRaw = r['id'];
      if (idRaw is RecordId) return idRaw;
      return parseRecordIdString(idRaw.toString());
    }).toList();

    if (idsToCheck.isEmpty) return;

    try {
      final resStr = await remote.query(
        r'SELECT id FROM $ids',
        vars: {'ids': idsToCheck},
      );
      final dynamic raw = extractResult(resStr);

      final existingIds = <String>{};
      if (raw is List) {
        for (final item in raw) {
          existingIds.add(item['id'].toString());
        }
      }

      final toDelete = <RecordId>[];
      for (final id in idsToCheck) {
        if (id is RecordId && !existingIds.contains(id.toString())) {
          toDelete.add(id);
        }
      }

      if (toDelete.isNotEmpty) {
        await local.query(r'DELETE $ids', vars: {'ids': toDelete});
        print('[QueryManager] Purged ${toDelete.length} orphaned records.');
      }
    } catch (e) {
      print('[QueryManager] Failed to purge orphans: $e');
    }
  }

  Future<void> _startLiveQuery() async {
    remote.subscribeLive(
      tableName: '_spooky_incantation',
      callback: (action, result) {
        if (action == 'UPDATE' || action == 'CREATE') {
          final id = result['id'] as String?;
          final hash = result['Hash'] as String?;
          final tree = result['Tree'];

          if (id == null || hash == null) return;

          final parts = id.split(':');
          final queryHash = parts.length > 1 ? parts[1] : parts[0];

          final incantation = activeQueries[queryHash];
          if (incantation == null) return;

          events.addEvent(
            IncantationRemoteHashUpdate(
              IncantationRemoteHashUpdatePayload(
                incantationId: id,
                surrealql: incantation.surrealql,
                localHash: incantation.hash,
                localTree: incantation.tree,
                remoteHash: hash,
                remoteTree: tree,
              ),
            ),
          );
        }
      },
    );
  }

  Future<String> _calculateHash(
    String surrealql,
    Map<String, dynamic> params,
  ) async {
    final content = jsonEncode({'surrealql': surrealql, 'params': params});

    try {
      final resultJson = await local.query(
        r'RETURN crypto::blake3($content)',
        vars: {'content': content},
      );

      final rawList = jsonDecode(resultJson);
      if (rawList is List && rawList.isNotEmpty) {
        final val = extractResult(resultJson);
        return val.toString();
      }
    } catch (e) {
      print(
        "[QueryManager] Hash calc error: $e. Using simple string hash fallback.",
      );
    }

    return content.hashCode.toString();
  }
}
